java/ql/src/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -138,127 +188,420 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ";CollectionUtils;true;addAll;(Collection,Enumeration);;Element of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addAll;(Collection,Iterable);;Element of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addAll;(Collection,Iterator);;Element of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addIgnoreNull;;;Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;collate;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;collate;;;Element of Argument[1];Element of ReturnValue;value",
+          ";CollectionUtils;true;disjunction;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;disjunction;;;Element of Argument[1];Element of ReturnValue;value",
+          ";CollectionUtils;true;emptyIfNull;;;Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;extractSingleton;;;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;find;;;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Iterator,int);;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Iterable,int);;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Map,int);;MapKey of Argument[0];MapKey of ReturnValue;value",
+          ";CollectionUtils;true;get;(Map,int);;MapValue of Argument[0];MapValue of ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;ArrayElement of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;MapKey of Argument[0];MapKey of ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;MapValue of Argument[0];MapValue of ReturnValue;value",
+          ";CollectionUtils;true;getCardinalityMap;;;Element of Argument[0];MapKey of ReturnValue;value",
+          ";CollectionUtils;true;intersection;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;intersection;;;Element of Argument[1];Element of ReturnValue;value",
+          ";CollectionUtils;true;permutations;;;Element of Argument[0];Element of Element of ReturnValue;value",
+          ";CollectionUtils;true;predicatedCollection;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;removeAll;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;retainAll;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;select;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;select;;;Element of Argument[0];Element of Argument[2];value",
+          ";CollectionUtils;true;select;;;Element of Argument[0];Element of Argument[3];value",
```

Comment:
Note these two also return their `Argument[2]` (distinct from the element-transfer to return value case which only applies to the one-arg overload)

Reply:
Yes, the `Iterable, Predicate` version should be modelled elem-of-arg0 -> elem-of-ret, whereas the `Iterable, Predicate, Collection[, Collection]` overloads should be modelled elem-of-arg0 -> elem-of-arg2 and arg2 -> ret (cf elem of arg(0 or 2) -> elem of ret which would obscure the fact that arg2 and ret are the same object)

Reply:
I was worried that we wouldn't get elem-of-arg0 -> elem-of-ret because "arg2 -> ret" would consider "arg2" from before this line is executed. But I just tested it and it works fine 😄 .